nexus/src/app/background/driver.rs

```diff
@@ -54,9 +54,9 @@ struct Task {
     status: watch::Receiver<TaskStatus>,
     /// join handle for the tokio task that's executing this background task
     tokio_task: tokio::task::JoinHandle<()>,
-    /// `Notify` used to wake up the tokio task when a caller explicit wants to
+    /// `Activator` used to wake up the tokio task when a caller explicit wants to
     /// activate the background task
-    notify: Arc<Notify>,
+    activator: Activator,
```

Comment:
I'm not totally clear on why we need to store the `Activator` here.  I find this a little bit confusing because I think of an Activator as a handle we pass _outside_ this subsystem to support a one-directional message: all you can do with this thing is activate it, whatever that means.  Now, it's bidirectional: you can wait on activation, too.  And it's both an internal and external interface.  To be clear there's nothing really wrong with this but I find it a bit more confusing so that's why I'm asking to better understand why.

The new `debug_assert!` is pretty good though and I think that's only possible because of this.

Reply:
Yeah, I was also a bit on the fence about this part. 

The driver could, alternatively, own an `Arc<ActivatorInner>`, without the `Activator` newtype around it, so it would no longer have access to the `activate()` method, and `activated()` would no longer be a private API defined on `Activator`. But, the driver would in that case still have access to the `Notify` inside the `Activator`, which it could use to do either of those things anyway, in theory. So, I'm not sure if unwrapping it actually meaningfully changes anything.

We could also remove the `Activator::activated()` method and just call `activator.0.notify.notified()`, but that way, we are no longer ensuring that the assertion is made every time we await an activation. Since there's only one place where we await an activation, though, that's not _currently_ a problem...but this does future-proof it a bit.

I'm happy to go either way. I also considered adding a private newtype that represents the "await activation" side, but that seemed like overengineering it a bit to separate out an interface that's totally internal to this module...